File: enaible-ts/src/commands/prompts.ts
/**
 * Prompt command group for Enaible CLI.
 */

import { Command } from 'commander';
import path from 'path';
import fs from 'fs';
import chalk from 'chalk';
import { glob } from 'glob';
import { loadWorkspace } from '../runtime/context.js';
import { PromptRenderer, writeRenderResult, getRenderResultDiff } from '../prompts/renderer.js';
import { CATALOG } from '../prompts/catalog.js';
import { lintFiles } from '../prompts/lint.js';

const GENERATED_SENTINEL = '<!-- generated: enaible -->';

/**
 * Resolve prompt IDs from input.
 */
function resolvePromptIds(prompts: string[]): string[] {
  if (!prompts.length || prompts.includes('all')) {
    return Object.keys(CATALOG);
  }

  const catalogIds = new Set(Object.keys(CATALOG));
  const unknown = prompts.filter((p) => !catalogIds.has(p));
  if (unknown.length > 0) {
    const available = Array.from(catalogIds).sort().join(', ');
    throw new Error(`Unknown prompt(s): ${unknown.join(', ')}. Available: ${available}`);
  }
  return prompts;
}

/**
 * Resolve systems from input.
 */
function resolveSystems(promptIds: string[], systems: string[]): string[] {
  if (!systems.length || systems.includes('all')) {
    const supported = new Set<string>();
    for (const promptId of promptIds) {
      const definition = CATALOG[promptId];
      for (const system of Object.keys(definition.systems)) {
        supported.add(system);
      }
    }
    return Array.from(supported).sort();
  }
  return systems;
}

/**
 * Build output overrides map.
 */
function buildOverrides(
  selectedSystems: string[],
  out: string | null
): Record<string, string | null> {
  const overrides: Record<string, string | null> = {};
  if (!out) {
    for (const system of selectedSystems) {
      overrides[system] = null;
    }
    return overrides;
  }

  if (selectedSystems.length === 1) {
    overrides[selectedSystems[0]] = out;
    return overrides;
  }

  for (const system of selectedSystems) {
    overrides[system] = path.join(out, system);
  }
  return overrides;
}

/**
 * Split comma-separated values.
 */
function splitCsv(value: string): string[] {
  if (!value || value.toLowerCase() === 'all') {
    return ['all'];
  }
  return value
    .split(',')
    .map((s) => s.trim())
    .filter((s) => s);
}

/**
 * Create prompts command.
 */
export function createPromptsCommand(): Command {
  const prompts = new Command('prompts').description('Render and validate managed prompts.');

  // List command
  prompts
    .command('list')
    .description('List prompts known to the catalog.')
    .action(() => {
      const context = loadWorkspace();
      const renderer = new PromptRenderer(context);
      for (const definition of renderer.listPrompts()) {
        const systems = Object.keys(definition.systems).sort().join(', ');
        console.log(`${definition.promptId}: ${definition.title} [${systems}]`);
      }
    });

  // Render command
  prompts
    .command('render')
    .description('Render prompts for the selected systems.')
    .option('--prompt <prompts>', "Comma-separated prompt identifiers or 'all'.", 'all')
    .option('--system <systems>', "Comma-separated system identifiers or 'all'.", 'all')
    .option('-o, --out <path>', 'Optional override directory for rendered output.')
    .action((options) => {
      const context = loadWorkspace();
      const renderer = new PromptRenderer(context);

      const promptArgs = splitCsv(options.prompt);
      const systemArgs = splitCsv(options.system);

      const selectedPrompts = resolvePromptIds(promptArgs);
      const selectedSystems = resolveSystems(selectedPrompts, systemArgs);
      const overrides = buildOverrides(selectedSystems, options.out ?? null);

      const results = renderer.render(selectedPrompts, selectedSystems, overrides);

      for (const result of results) {
        writeRenderResult(result);
        console.log(
          `Rendered ${chalk.cyan(result.promptId)} for ${chalk.yellow(result.system)} → ${result.outputPath}`
        );
      }
    });

  // Diff command
  prompts
    .command('diff')
    .description('Show diffs between catalog output and current files.')
    .option('--prompt <prompts>', "Comma-separated prompt identifiers or 'all'.", 'all')
    .option('--system <systems>', "Comma-separated system identifiers or 'all'.", 'all')
    .action((options) => {
      const context = loadWorkspace();
      const renderer = new PromptRenderer(context);

      const promptArgs = splitCsv(options.prompt);
      const systemArgs = splitCsv(options.system);

      const selectedPrompts = resolvePromptIds(promptArgs);
      const selectedSystems = resolveSystems(selectedPrompts, systemArgs);

      const results = renderer.render(selectedPrompts, selectedSystems);

      let hasDiff = false;
      for (const result of results) {
        const diffOutput = getRenderResultDiff(result);
        if (diffOutput) {
          hasDiff = true;
          console.log(diffOutput);
        }
      }

      if (hasDiff) {
        process.exit(1);
      }
    });

  // Validate command
  prompts
    .command('validate')
    .description('Validate that rendered prompts match committed files.')
    .option('--prompt <prompts>', "Comma-separated prompt identifiers or 'all'.", 'all')
    .option('--system <systems>', "Comma-separated system identifiers or 'all'.", 'all')
    .action((options) => {
      const context = loadWorkspace();
      const renderer = new PromptRenderer(context);

      const promptArgs = splitCsv(options.prompt);
      const systemArgs = splitCsv(options.system);

      const selectedPrompts = resolvePromptIds(promptArgs);
      const selectedSystems = resolveSystems(selectedPrompts, systemArgs);

      const results = renderer.render(selectedPrompts, selectedSystems);

      let hasDiff = false;
      for (const result of results) {
        const diffOutput = getRenderResultDiff(result);
        if (diffOutput) {
          hasDiff = true;
          console.log(diffOutput);
        }
      }

      if (hasDiff) {
        console.log(chalk.red('Prompt drift detected. Run `enaible prompts render` to update.'));
        process.exit(1);
      }
    });

  // Lint command
  prompts
    .command('lint')
    .description('Lint prompt sources for @TOKEN usage and variable mapping rules.')
    .option('--prompt <prompts>', "Comma-separated prompt identifiers or 'all'.", 'all')
    .action(async (options) => {
      const context = loadWorkspace();

      const selectedPrompts = resolvePromptIds(splitCsv(options.prompt));

      // Collect unique source files for selected prompts
      const files = new Set<string>();
      for (const promptId of selectedPrompts) {
        const definition = CATALOG[promptId];
        const sourcePath = path.resolve(context.repoRoot, definition.sourcePath);
        files.add(sourcePath);
      }

      // Also lint unmanaged, hand-authored system prompts
      const systemDirs = [
        path.join(context.repoRoot, 'systems', 'claude-code', 'commands'),
        path.join(context.repoRoot, 'systems', 'codex', 'prompts'),
        path.join(context.repoRoot, 'systems', 'copilot', 'prompts'),
        path.join(context.repoRoot, 'systems', 'cursor', 'rules'),
      ];

      for (const dir of systemDirs) {
        if (!fs.existsSync(dir)) continue;
        const mdFiles = await glob('*.md', { cwd: dir });
        for (const mdFile of mdFiles) {
          if (mdFile.toLowerCase() === 'agents.md') continue;
          const fullPath = path.join(dir, mdFile);
          try {
            const head = fs.readFileSync(fullPath, 'utf-8').split('\n').slice(0, 3);
            if (head.some((line) => line.includes(GENERATED_SENTINEL))) {
              continue;
            }
            files.add(fullPath);
          } catch {
            continue;
          }
        }
      }

      const issues = lintFiles(Array.from(files));
      if (!issues.length) {
        console.log(chalk.green('prompts: lint passed'));
        return;
      }

      for (const issue of issues) {
        console.log(`${issue.path}:${issue.line}: ${issue.message}`);
      }
      process.exit(1);
    });

  return prompts;
}

File: ai-assisted-workflows/tools/enaible-rs/crates/enaible-cli/src/commands/prompts.rs
use anyhow::Result;
use clap::{Args, Subcommand};
use enaible_core::load_workspace;
use enaible_prompts::{lint_files, split_csv, PromptRenderer, CATALOG};
use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};

#[derive(Subcommand)]
pub enum PromptsCommands {
    /// List prompts known to the catalog
    List,

    /// Render prompts for the selected systems
    Render {
        /// Comma-separated prompt identifiers or 'all'
        #[arg(long, default_value = "all")]
        prompt: String,

        /// Comma-separated system identifiers or 'all'
        #[arg(long, default_value = "all")]
        system: String,

        /// Optional override directory for rendered output
        #[arg(short, long)]
        out: Option<PathBuf>,
    },

    /// Show diffs between catalog output and current files
    Diff {
        /// Comma-separated prompt identifiers or 'all'
        #[arg(long, default_value = "all")]
        prompt: String,

        /// Comma-separated system identifiers or 'all'
        #[arg(long, default_value = "all")]
        system: String,
    },

    /// Validate that rendered prompts match committed files
    Validate {
        /// Comma-separated prompt identifiers or 'all'
        #[arg(long, default_value = "all")]
        prompt: String,

        /// Comma-separated system identifiers or 'all'
        #[arg(long, default_value = "all")]
        system: String,
    },

    /// Lint prompt sources for @TOKEN usage and variable mapping rules
    Lint {
        /// Comma-separated prompt identifiers or 'all'
        #[arg(long, default_value = "all")]
        prompt: String,
    },
}

pub fn handle_command(cmd: PromptsCommands) -> Result<()> {
    match cmd {
        PromptsCommands::List => prompts_list(),
        PromptsCommands::Render { prompt, system, out } => {
            prompts_render(&prompt, &system, out)
        }
        PromptsCommands::Diff { prompt, system } => prompts_diff(&prompt, &system),
        PromptsCommands::Validate { prompt, system } => prompts_validate(&prompt, &system),
        PromptsCommands::Lint { prompt } => prompts_lint(&prompt),
    }
}

fn prompts_list() -> Result<()> {
    let context = load_workspace(None)?;
    let renderer = PromptRenderer::new(context)?;

    for definition in renderer.list_prompts() {
        let systems: Vec<String> = definition.systems.keys().cloned().collect();
        println!(
            "{}: {} [{}]",
            definition.prompt_id,
            definition.title,
            systems.join(", ")
        );
    }

    Ok(())
}

fn resolve_prompt_ids(prompts: &[String]) -> Result<Vec<String>> {
    if prompts.is_empty() || prompts == ["all"] {
        return Ok(CATALOG.keys().cloned().collect());
    }

    let catalog_ids: HashSet<_> = CATALOG.keys().cloned().collect();
    let unknown: Vec<_> = prompts
        .iter()
        .filter(|p| !catalog_ids.contains(*p))
        .collect();

    if !unknown.is_empty() {
        let available: Vec<_> = catalog_ids.into_iter().collect();
        anyhow::bail!(
            "Unknown prompt(s): {}. Available: {}",
            unknown
                .into_iter()
                .map(|s| s.as_str())
                .collect::<Vec<_>>()
                .join(", "),
            available.join(", ")
        );
    }

    Ok(prompts.to_vec())
}

fn resolve_systems(prompt_ids: &[String], systems: &[String]) -> Vec<String> {
    if systems.is_empty() || systems == ["all"] {
        let mut supported = HashSet::new();
        for prompt_id in prompt_ids {
            if let Some(definition) = CATALOG.get(prompt_id) {
                supported.extend(definition.systems.keys().cloned());
            }
        }
        return supported.into_iter().collect();
    }

    systems.to_vec()
}

fn build_overrides(
    selected_systems: &[String],
    out: Option<PathBuf>,
) -> HashMap<String, Option<PathBuf>> {
    let mut overrides = HashMap::new();

    if let Some(out) = out {
        if selected_systems.len() == 1 {
            overrides.insert(selected_systems[0].clone(), Some(out));
        } else {
            for system in selected_systems {
                overrides.insert(system.clone(), Some(out.join(system)));
            }
        }
    } else {
        for system in selected_systems {
            overrides.insert(system.clone(), None);
        }
    }

    overrides
}

fn prompts_render(prompts: &str, systems: &str, out: Option<PathBuf>) -> Result<()> {
    let context = load_workspace(None)?;
    let renderer = PromptRenderer::new(context)?;

    let prompt_args = split_csv(prompts);
    let system_args = split_csv(systems);

    let selected_prompts = resolve_prompt_ids(&prompt_args)?;
    let selected_systems = resolve_systems(&selected_prompts, &system_args);
    let overrides = build_overrides(&selected_systems, out);

    let results = renderer.render(&selected_prompts, &selected_systems, Some(overrides))?;

    for result in results {
        result.write()?;
        println!(
            "Rendered {} for {} → {}",
            result.prompt_id,
            result.system,
            result.output_path.display()
        );
    }

    Ok(())
}

fn prompts_diff(prompts: &str, systems: &str) -> Result<()> {
    let context = load_workspace(None)?;
    let renderer = PromptRenderer::new(context)?;

    let prompt_args = split_csv(prompts);
    let system_args = split_csv(systems);

    let selected_prompts = resolve_prompt_ids(&prompt_args)?;
    let selected_systems = resolve_systems(&selected_prompts, &system_args);

    let results = renderer.render(&selected_prompts, &selected_systems, None)?;

    let mut has_diff = false;
    for result in results {
        let diff_output = result.diff()?;
        if !diff_output.is_empty() {
            has_diff = true;
            println!("{}", diff_output);
        }
    }

    if has_diff {
        std::process::exit(1);
    }

    Ok(())
}

fn prompts_validate(prompts: &str, systems: &str) -> Result<()> {
    match prompts_diff(prompts, systems) {
        Ok(_) => Ok(()),
        Err(_) => {
            eprintln!("Prompt drift detected. Run `enaible prompts render` to update.");
            std::process::exit(1);
        }
    }
}

fn prompts_lint(prompts: &str) -> Result<()> {
    let context = load_workspace(None)?;

    let prompt_args = split_csv(prompts);
    let selected_prompts = resolve_prompt_ids(&prompt_args)?;

    // Collect unique source files for selected prompts
    let mut files = HashSet::new();
    for prompt_id in &selected_prompts {
        if let Some(definition) = CATALOG.get(prompt_id) {
            let path = context.repo_root.join(&definition.source_path);
            files.insert(path);
        }
    }

    // Also lint unmanaged, hand-authored system prompts
    let generated_sentinel = "<!-- generated: enaible -->";
    let system_dirs = [
        context.repo_root.join("systems/claude-code/commands"),
        context.repo_root.join("systems/codex/prompts"),
        context.repo_root.join("systems/copilot/prompts"),
        context.repo_root.join("systems/cursor/rules"),
    ];

    for dir in &system_dirs {
        if !dir.exists() {
            continue;
        }

        for entry in std::fs::read_dir(dir)? {
            let entry = entry?;
            let path = entry.path();

            if path.extension().and_then(|s| s.to_str()) == Some("md") {
                let name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
                if name.to_lowercase() == "agents.md" {
                    continue;
                }

                if let Ok(content) = std::fs::read_to_string(&path) {
                    let head: Vec<_> = content.lines().take(3).collect();
                    if head.iter().any(|line| line.contains(generated_sentinel)) {
                        continue;
                    }
                    files.insert(path);
                }
            }
        }
    }

    let issues = lint_files(&files)?;
    if issues.is_empty() {
        println!("prompts: lint passed");
    } else {
        for issue in &issues {
            println!("{}:{}: {}", issue.path, issue.line, issue.message);
        }
        std::process::exit(1);
    }

    Ok(())
}